src/buildstream_plugins/sources/cargo.py

```diff
@@ -294,6 +308,166 @@ def _get_mirror_file(self, sha=None):
         return os.path.join(self._get_mirror_dir(), sha or self.sha)
 
 
+# Locks on repositories for write access
+REPO_LOCKS = {}
+
+
+# CrateGit()
+#
+# Use a SourceFetcher class to be the per crate helper.
+#
+# This one is for crates fetched from git repositories.
+#
+# Args:
+#    cargo (Cargo): The main Source implementation
+#    name (str): The name of the crate to depend on
+#    version (str): The version of the crate to depend on
+#    repo (str): Repository URL
+#    commit (str): Sha of the git commit
+class CrateGit(SourceFetcher):
+    def __init__(self, cargo, name, version, repo, commit):
+        super().__init__()
+
+        self.cargo = cargo
+        self.name = name
+        self.version = str(version)
+        self.repo = repo
+        self.commit = commit
+        # TODO: Is this right?
+        self.mark_download_url(self.repo)
```

Comment:
I'm a little bit confused what this is for. I think the other fetcher just passes "cargo:cargo"? So this needs someone who understands this.

Reply:
Should I implement this kind of mirror code?

Reply:
@sophie-h if you want to look at real-world example, https://gitlab.gnome.org/GNOME/gnome-build-meta/-/blob/master/include/aliases.yml#L12 -> https://gitlab.gnome.org/GNOME/gnome-build-meta/-/blob/master/include/mirrors.yml#L10. You as plugin author only need to provide ability to provide aliased URL's. BuildStream expands aliases based on user (or project) configuration and this may include mirrors. freedesktop-sdk and gnome attempt to mirror everything through lorry.

Reply:
> if you want to look at real-world example

I don't see how this is related. The cargo plugin generates the list of dependencies when calling `source track` from the `Cargo.lock` file. There won't be any aliases used in the crate refs (dependencies).

As I understand @AdrianVovk's suggestion, this would change all git pulls to be redirected to a single git host, independent of the actual git host of the original URL. The aliasing mechanism would only be used for this one mirror URL. So that's something different I guess?

Reply:
Aliasing indeed needs to be set on per-git repo level.

Reply:
> Aliasing indeed needs to be set on per-git repo level.

Okay, so we just go through all git sources in `configure` and call `mark_download_url` for them and then just call `translate_url` on the complete URL. That's then the same as the `git_repo` plugin does, just that we have to potentially deal with multiple repos per `Source` object.

Sounds easy enough.

Reply:
Okay, that's not working since buildstream doesn't seem to have a feature to translate one URL to another. The only thing that's supported is adding an explicit prefix to a URL, which we can't do because the URLs are auto-generated by the plugin.

If that's correct, the only solution I can see is to add a list of URL mappings to the source config like

```yaml
sources:
- kind: cargo
  git_mirrors:
  - https://github.com/x/y: https://mirror.example.org/x/y
  - https://github.com/a/b: my-alias:a/b
```

Reply:
Yeah, sounds legit. @abderrahim and @gtristan were reworking how URL marking works but it doesn't seem it helps here.

Reply:
Please consider the fact that rust projects have _lots_ of dependencies...

I'd rather not have to configure the alias mappings for dozens of rust crates each time I want to use the cargo plugin

Counterpoint: I guess rust projects don't have that many git dependencies necessarily, and we could configure the URL mapping globally, project-wide to avoid duplicating work?

Just something to keep in mind

Reply:
FWIW: The typical case in GNOME would probably be that git dependencies pop up and disappear again because the project wants to test out some other components that don't have a release yet.

Reply:
> ```yaml
> sources:
> - kind: cargo
>   git_mirrors:
>   - https://github.com/x/y: https://mirror.example.org/x/y
>   - https://github.com/a/b: my-alias:a/b
> ```

What I have done there is not what mirrors do in practice. See this mirror entry:

```yaml
    github:
    - https://gitlab.gnome.org/Infrastructure/Mirrors/lorry-mirrors/github_com/
```

Maybe what's called `git_mirrors` above could be a string replace list such that

> ```yaml
> sources:
> - kind: cargo
>   git_mirrors:
>   - https://github.com/: github:
> ```

works too?

Reply:
That would allow us to mirror a specific repo or to just have a complete GitHub mirror.

Reply:
:+1: ok I like it.

Small concern; it should behave like keys. That way projects can configure a bunch of project-wide `git_mirrors` so it doesn't have to be specified in each element, but also if you _do_ need to configure something per-element you can just configure it

With your proposal right now, the set-up would need to look like:
```
sources:
- kind: cargo
  git_mirrors:
    (>):
    - https://github.com/x/y: https://mirror.example.org/x/y
    - https://github.com/a/b: my-alias:a/b
```

to avoid overwriting the project-wide config

Of course, domains don't fit into YAML syntax, so maybe what I'm asking for isn't possible :shrug:

Reply:
Oh, there is a consistent typo. URLs actually do work as keys so it should be

```yaml
  git_mirrors:
   https://github.com/x/y: 
    - https://mirror.example.org/x/y
   https://github.com/a/b:
    - my-alias:a/b
```

I guess I will try to implement this then next week

Reply:
Should probably be like this then

```yaml
git_mirrors:
   https://github.com/x/y: https://mirror.example.org/x/y
   https://github.com/a/b: my-alias:a/b
```

Reply:
Okay, implemented this and tried to also document it.